### ISSUE TITLE
__hadd() is ambiguous when EIGEN_CUDA_ARCH >= 530

### ISSUE THREAD

**Issue by chaosink** [tensorflow/tensorflow#19198]
### System information
- **Have I written custom code**: I change the CUDA capabilities to 6.1 and 7.0.
- **OS Platform and Distribution**: Windows 10
- **TensorFlow installed from (source or binary)**: I'm compiling the source code.
- **TensorFlow version (use command below)**: branch `r1.8`, 8753e2ebde6c58b56675cc19ab7ff83072824a62
- **Python version**: 3.6.0
- **Bazel version (if compiling from source)**: No
- **GCC/Compiler version (if compiling from source)**: VS 2017(v141), but v140 for CUDA host compiler
- **CUDA/cuDNN version**: CUDA 9.0, cuDNN 7.1
- **GPU model and memory**: 1080 Ti, Titan V
- **Exact command to reproduce**: cmake-gui, enable GPU, and change CUDA host compiler to v140

### Describe the problem
[__hadd()](https://github.com/eigenteam/eigen-git-mirror/blob/2bdd9e80c43ee491fb0cb299940995e094b1647b/Eigen/src/Core/arch/CUDA/Half.h#L212) is ambiguous when [EIGEN_CUDA_ARCH >= 530](https://github.com/eigenteam/eigen-git-mirror/blob/2bdd9e80c43ee491fb0cb299940995e094b1647b/Eigen/src/Core/arch/CUDA/Half.h#L204).

The following is where the ambiguity comes from found in VS 2017:
![image](https://user-images.githubusercontent.com/6009211/39873593-f19954bc-549d-11e8-8bb6-3c02e33c44c5.png)

### Source code / logs
`tf_core_gpu_kernels` compilation fails because of this problem:
```
42>Building NVCC (Device) object CMakeFiles/tf_core_gpu_kernels.dir/__/__/core/kernels/Release/tf_core_gpu_kernels_generated_check_numerics_op_gpu.cu.cc.obj
42>check_numerics_op_gpu.cu.cc
42>e:\program\ml\tensorflow_build_05-10-01\external\eigen_archive\eigen\src/Core/arch/CUDA/Half.h(212): error : more than one instance of overloaded function "__hadd" matches the argument list:
42>            function "__hadd(int, int)"
42>            function "__hadd(__half, __half)"
42>            argument types are: (const Eigen::half, const Eigen::half)
```

I'm confused that nobody has every post such an issue.
Nobody has ever tried changing CUDA capabilities to >=5.3?
Or is there something wrong with my environment?

It seems that this is a pure Eigen issue...

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
@rmlarsen @benoitsteiner : Mind taking a look?

**Comment by rmlarsen** [comment]
@asimshankar thanks, we should apply the patch upstream.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @benoitsteiner, @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @benoitsteiner, @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bstriner** [comment]
I just submitted a PR so you can do this kind of patching cleanly. Added a flag `eigen_PATCH_FILE` to cmake build so you can apply the patch automatically at the right time in the build. Also linked a patch file there based off the one in the bug report that works on the current version of eigen in tensorflow.

Not that this shouldn't be fixed upstream, but it is nice to have a little flexibility with patching as well.

https://github.com/tensorflow/tensorflow/pull/21119

**Comment by ruanjiandong** [comment]
What configuration is used to build the official Windows build? With the recent change (commit aef000ed3c 10 days ago), the official Windows build should break with similar issue.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @benoitsteiner, @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by trias702** [comment]
Is there any way of integrating the Eigen fix into the r1.10 tensorflow build using Bazel? This is currently breaking my Windows build, and I'm not sure how to slipstream the Eigen patch file into the Bazel build.

Is there any way to have r1.10 download a different version of Eigen during the build process?

**Comment by ruanjiandong** [comment]
add the following line to tensorflow/workspace.bzl for eigen_archive. Copy patch file to third_party directory.
        patch_file = clean_dep("//third_party:eigen_half.patch"),

**Comment by trias702** [comment]
@ruanjiandong Many thanks for the reply!

If I follow your instructions, can I use the patch file I download from here, which fixes this issue: http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1526

Or do I need to re-write that patch file to a different format which Bazel will understand?

**Comment by ruanjiandong** [comment]
The patch file in eigen site won't match tensorflow 1.10. For tensorflow 1.10, you can use patch file https://gist.github.com/bstriner/a7fb0a8da1f830900fa932652439ed44

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by trias702** [comment]
That would work well for me.

I'm currently building with CMake using a hack which upstreams the patch,
also did a similar thing for Bazel.

But yeah, if you can amend TF master to do it automatically for
Bazel/Cmake, that would be great.

**Comment by ruanjiandong** [comment]
@gunan , I've submitted eigen pull request. For status, see https://bitbucket.org/eigen/eigen/pull-requests/481

**Comment by ruanjiandong** [comment]
@gunan eigen fix is merged into main development branch. The download link is https://bitbucket.org/eigen/eigen/get/52517c8764ad.tar.gz

**Comment by yongtang** [comment]
I tried to apply the changes:
https://github.com/tensorflow/tensorflow/compare/master...yongtang:19198-eigen

Though it looks like there are build failures. Will take a look and see if the build failure could be fixed.

**Comment by gunan** [comment]
Thanks for upstreaming the change into eigen.
I learned that our upgrade of eigen is blocked right now due to internally and externally a few eigen pieces being out of sync.
@rmlarsen is right now working to get this resolved. We will then upgrade our eigen dependency once all issues are resolved. However, the current issue is a rarther involved one and it may take up to a couple of weeks.

**Comment by ruanjiandong** [comment]
We need one more fix in eigen to fix the break introduced by the new eigen "block" feature. I created a pull request in eigen repo and reply here once it is merged.

The remaining changes (10+ files) can be done tensorflow repo. They're mostly for renaming symbols. @rmlarsen , I can make the tensorflow change if you haven't started.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gunan** [comment]
I think eigen upgrade is complete.
will close this bug.

**Comment by ItWasAllIntended** [comment]
What commit was Eigen bumped in? Not in any of the releases at the moment yes?

**Comment by gunan** [comment]
1.12 did not have the eigen version upgrade.
I do not think we will backport the fix to 1.12, as it was quite a large and complicated change to incorporate.

**Comment by VCBE123** [comment]
> > I think eigen upgrade is complete.
> > will close this bug.
> 
> @gunan In fact even for `Tensorflow v1.12.0` and `Bazel 0.15.0` on Windows the bug still exists when building GPU version.

so, how to solve this problem

**Comment by gunan** [comment]
Please see https://github.com/tensorflow/tensorflow/issues/19198#issuecomment-456508088

At the moment, 1.12 branch is more than 1 year old.
Therefore, we have completely dropped support for that version.
Please use a newer branch.
